combine faq and best practices readme/markdown docs into one doc as faq doc. (#917)
commercetools/commercetools-adyen-integration: docs/FAQ.md

- An encrypted HTTPS connection strongly recommended for production setups instead of HTTP connection.
- To protect your extension service from unauthorized calls, we recommend to [activate basic auth authentication](../extension/docs/HowToRun.md#commercetools) on the API extension to prevent unauthorized 3rd-party from accessing extension endpoint.
- To protect your notification service from unauthorized notifications, we strongly recommend that you activate Hash-based message authentication code [HMAC signatures](../notification/docs/IntegrationGuide.md#step-1-set-up-notification-webhook-and-generate-hmac-signature) during the notification setup.

### What is the best practices for creating/maintaining commercetools payment object ?

- Create your commercetools payment with following the [creating commercetools payment guide](../extension/docs/WebComponentsIntegrationGuide.md#step-2-creating-a-commercetools-payment).
- After successful payment creation always [add](https://docs.commercetools.com/api/projects/carts#add-payment) it to the appropriate cart.
- **Never delete or un-assign** created payment objects during the checkout from the cart.

> - Currently [payment status](https://docs.commercetools.com/api/projects/payments#paymentstatus) is not maintained by the integration as status of the payment can be derived from the state of the payment transaction(s).

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/917/commits/9f22b52dfb1d2d595e7a9286781e41c9ebfde520

### Is there any error handling mechanism required while using commercetools-adyen-integration ?

Yes, in case you encounter errors in your integration, refer to the [error-handling section](../extension/docs/WebComponentsIntegrationGuide.md#error-handling) on the integration documentation.

### When we should create commercetools order ? 

There are 2 approaches for creating commercetools order, one is creating order after a successful payment, and other is creating before the successful payment.
Both approaches have their good and bad sides, but we found out that creating an order AFTER a successful payment is less harmful. Please read the details below for both approaches and decide based on your checkout scenarios.

#### What to consider when creating order _AFTER_ a successful commercetools payment ?

> - **Shop was not reachable due to network issues**: It is possible that shop was not reachable due to network issues. In that case there will be a successful payment but no order.

Reviewer: ```suggestion
- **Shop (success  redirect URL) is not reachable due to the network issues**: Since shop creates an order and  success shop redirect URL can not be reached we might end up with a successful payment but no order.
```

>   - **Possible solution**: Create an order based on the payment transaction changes, which delivery is guaranteed due to the asynchronous notifications from Adyen. Depending on your preference you might either query for the latest messages of type [PaymentTransactionAdded](https://docs.commercetools.com/api/message-types#paymenttransactionadded-message), [PaymentTransactionStateChanged](https://docs.commercetools.com/api/message-types#paymenttransactionstatechanged-message) or [subscribe](https://docs.commercetools.com/api/projects/subscriptions#create-a-subscription) to the mentioned message types. Every [message](https://docs.commercetools.com/api/message-types#message) will link its payment through the `resource` field and since every payment is attached to a cart one has all the informations at hand to decide if the cart has to be converted to order or not. Since the job or worker processing the message is not a usual place where the order is created it might be reasonable to pass the cart ID to another service or web shop URL which will verify the cart and create an order out of it.

Reviewer: ```suggestion
  - **Possible solution**: Create an order asynchronously based on the payment transaction changes, which delivery is guaranteed due to the asynchronous notifications from Adyen. Depending on your preference you might either query for the latest messages of type [PaymentTransactionAdded](https://docs.commercetools.com/api/message-types#paymenttransactionadded-message), [PaymentTransactionStateChanged](https://docs.commercetools.com/api/message-types#paymenttransactionstatechanged-message) or [subscribe](https://docs.commercetools.com/api/projects/subscriptions#create-a-subscription) to the mentioned message types. Every [message](https://docs.commercetools.com/api/message-types#message) will link its payment through the `resource` field and since every payment is attached to a cart one has all the informations at hand to decide if the cart has to be converted to order or not. Since the job or worker processing the message is not a usual place where the order is created it might be reasonable to pass the cart ID to another service or web shop URL which will verify the cart and create an order out of it.
```

> - **More than 1 valid payments on the cart/order**: It is possible to have more than 1 valid payments on the cart/order. During checkout in two tabs, two redirect payments

Reviewer: ```suggestion
- **More than 1 successful payments on the cart/order**: It is possible to have more than 1 valid payments on the cart/order. It could happen for example when customer initiates a payment in two different tabs for the same cart and both payments are of type redirect (like credit card and paypal.
```

>   (like credit card and paypal) can be created for the same cart. These two payments can be completed independently in both tabs. Since every payment should be always attached to the cart this would make a cart to link two successful payments.

Reviewer: ```suggestion
 These two payments can be completed independently in both tabs. Since every payment should be always attached to the cart this would make a cart to link two successful payments.
```

  - **Possible solution**: [Refund](https://github.com/commercetools/commercetools-adyen-integration/blob/master/extension/docs/Refund.md) one of the successful payments. Similar as in case of `Create an order based on transaction state changes` above one could process the same message types in order to figure out if the cart has to many successful payments and create a refund.  

> - **The amount of the successful payment is lower than the cart amount**: During the checkout the shopper might navigate with two tabs, one with the cart and the other tab might be already a redirected payment with fixed amount like PayPal. Shopper can add more items to the cart so that the cart value is not equal to the  amount that will be paid in the payment provider tab.

Reviewer: ```suggestion
- **The amount of the successful payment is lower than the cart amount**: During the checkout the shopper might navigate with two tabs, one with the cart and the other tab might be already a redirected payment with fixed amount like PayPal. Shopper can add more items to the cart so that the cart value is not equal to the amount that will be paid in the payment provider tab.
```

>   - **Possible solution**: After a successful payment, validate the cart to check if the cart amount matches payment amount. If not, refund/cancel the payment and ask the shopper to pay again.

Reviewer: ```suggestion
  - **Possible solution**: After a successful payment, validate the cart to check if the cart amount matches the paid amount. If not, refund/cancel the payment and ask the shopper to pay again.
```
